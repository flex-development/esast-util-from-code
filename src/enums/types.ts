/**
 * @file Enums - types
 * @module esast-util-from-code/enums/types
 */

import type { EsastNode } from '@flex-development/esast'
import type { Type } from '@flex-development/unist-util-types'

/**
 * esast node types.
 *
 * @enum {Type<EsastNode>}
 */
enum types {
  ambientDeclaration = 'ambientDeclaration',
  argumentList = 'argumentList',
  arithmeticExpression = 'arithmeticExpression',
  arrayExpression = 'arrayExpression',
  arrayPattern = 'arrayPattern',
  arrayType = 'arrayType',
  arrowFunctionExpression = 'arrowFunctionExpression',
  asClause = 'asClause',
  asExpression = 'asExpression',
  assertionPredicate = 'assertionPredicate',
  assignmentExpression = 'assignmentExpression',
  assignmentPattern = 'assignmentPattern',
  awaitExpression = 'awaitExpression',
  bigint = 'bigint',
  bitwiseExpression = 'bitwiseExpression',
  blockStatement = 'blockStatement',
  boolean = 'boolean',
  breakStatement = 'breakStatement',
  callExpression = 'callExpression',
  callSignature = 'callSignature',
  catchClause = 'catchClause',
  classBody = 'classBody',
  classDeclaration = 'classDeclaration',
  classExpression = 'classExpression',
  classHeritage = 'classHeritage',
  comment = 'comment',
  computedExpression = 'computedExpression',
  conditionalExpression = 'conditionalExpression',
  conditionalType = 'conditionalType',
  constructSignature = 'constructSignature',
  constructorType = 'constructorType',
  continueStatement = 'continueStatement',
  debuggerStatement = 'debuggerStatement',
  decorator = 'decorator',
  decoratorList = 'decoratorList',
  directive = 'directive',
  doWhileStatement = 'doWhileStatement',
  elseStatement = 'elseStatement',
  emptyStatement = 'emptyStatement',
  enumBody = 'enumBody',
  enumDeclaration = 'enumDeclaration',
  enumMember = 'enumMember',
  equalityExpression = 'equalityExpression',
  exportDeclaration = 'exportDeclaration',
  exportSpecifier = 'exportSpecifier',
  exportSpecifiers = 'exportSpecifiers',
  expressionStatement = 'expressionStatement',
  extendsClause = 'extendsClause',
  fieldDefinition = 'fieldDefinition',
  finallyBlock = 'finallyBlock',
  forInStatement = 'forInStatement',
  forOfStatement = 'forOfStatement',
  forStatement = 'forStatement',
  fromClause = 'fromClause',
  functionDeclaration = 'functionDeclaration',
  functionExpression = 'functionExpression',
  functionSignature = 'functionSignature',
  functionType = 'functionType',
  genericType = 'genericType',
  identifier = 'identifier',
  ifStatement = 'ifStatement',
  implementsClause = 'implementsClause',
  importAssertion = 'importAssertion',
  importAttribute = 'importAttribute',
  importDeclaration = 'importDeclaration',
  importExpression = 'importExpression',
  importMeta = 'importMeta',
  importSpecifier = 'importSpecifier',
  importSpecifiers = 'importSpecifiers',
  indexSignature = 'indexSignature',
  indexedAccessType = 'indexedAccessType',
  inferType = 'inferType',
  interfaceBody = 'interfaceBody',
  interfaceDeclaration = 'interfaceDeclaration',
  intersectionType = 'intersectionType',
  labeledStatement = 'labeledStatement',
  logicalExpression = 'logicalExpression',
  mappedSignature = 'mappedSignature',
  mappedType = 'mappedType',
  memberExpression = 'memberExpression',
  methodDefinition = 'methodDefinition',
  methodSignature = 'methodSignature',
  modifier = 'modifier',
  modifierList = 'modifierList',
  moduleBody = 'moduleBody',
  moduleDeclaration = 'moduleDeclaration',
  moduleExpression = 'moduleExpression',
  namespaceDeclaration = 'namespaceDeclaration',
  namespaceExport = 'namespaceExport',
  namespaceImport = 'namespaceImport',
  newExpression = 'newExpression',
  nonNullExpression = 'nonNullExpression',
  nothing = 'nothing',
  null = 'null',
  number = 'number',
  objectExpression = 'objectExpression',
  objectLiteralType = 'objectLiteralType',
  objectPattern = 'objectPattern',
  parameter = 'parameter',
  parameterList = 'parameterList',
  parenthesizedExpression = 'parenthesizedExpression',
  parenthesizedType = 'parenthesizedType',
  property = 'property',
  propertySignature = 'propertySignature',
  regexp = 'regexp',
  relationalExpression = 'relationalExpression',
  restElement = 'restElement',
  restType = 'restType',
  returnStatement = 'returnStatement',
  root = 'root',
  satisfiesClause = 'satisfiesClause',
  satisfiesExpression = 'satisfiesExpression',
  sequenceExpression = 'sequenceExpression',
  spreadElement = 'spreadElement',
  staticBlock = 'staticBlock',
  staticBlockBody = 'staticBlockBody',
  string = 'string',
  super = 'super',
  switchBody = 'switchBody',
  switchCase = 'switchCase',
  switchStatement = 'switchStatement',
  taggedTemplateExpression = 'taggedTemplateExpression',
  templateElement = 'templateElement',
  templateLiteral = 'templateLiteral',
  templatePlaceholder = 'templatePlaceholder',
  this = 'this',
  throwStatement = 'throwStatement',
  tryStatement = 'tryStatement',
  tupleElement = 'tupleElement',
  tupleType = 'tupleType',
  typeAliasDeclaration = 'typeAliasDeclaration',
  typeAnnotation = 'typeAnnotation',
  typeArgumentList = 'typeArgumentList',
  typeAssertion = 'typeAssertion',
  typeAssertionExpression = 'typeAssertionExpression',
  typeKeyword = 'typeKeyword',
  typeParameter = 'typeParameter',
  typeParameterList = 'typeParameterList',
  typePredicate = 'typePredicate',
  unaryExpression = 'unaryExpression',
  unaryTypeExpression = 'unaryTypeExpression',
  undefined = 'undefined',
  unionType = 'unionType',
  updateExpression = 'updateExpression',
  variableDeclaration = 'variableDeclaration',
  variableDeclarator = 'variableDeclarator',
  whileStatement = 'whileStatement',
  withStatement = 'withStatement',
  yieldExpression = 'yieldExpression'
}

export default types
