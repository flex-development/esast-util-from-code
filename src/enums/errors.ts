/**
 * @file Enums - errors
 * @module esast-util-from-code/enums/errors
 */

/**
 * Error messages.
 *
 * @enum {string}
 */
enum errors {
  comment = 'Unterminated comment',
  string = 'Unterminated string constant'
}

export default errors
