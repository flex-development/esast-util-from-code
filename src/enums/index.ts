/**
 * @file Entry Point - Enums
 * @module esast-util-from-code/enums
 */

export { default as errors } from './errors'
export { default as tt } from './tt'
