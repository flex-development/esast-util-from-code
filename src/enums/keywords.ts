/**
 * @file Enums - keywords
 * @module esast-util-from-code/enums/keywords
 */

/**
 * Possible keywords.
 *
 * @enum {Lowercase<string>}
 */
enum keywords {
  abstract = 'abstract',
  accessor = 'accessor',
  any = 'any',
  as = 'as',
  assert = 'assert',
  asserts = 'asserts',
  async = 'async',
  await = 'await',
  bigint = 'bigint',
  boolean = 'boolean',
  break = 'break',
  case = 'case',
  catch = 'catch',
  class = 'class',
  const = 'const',
  constructor = 'constructor',
  continue = 'continue',
  debugger = 'debugger',
  declare = 'declare',
  default = 'default',
  delete = 'delete',
  do = 'do',
  else = 'else',
  enum = 'enum',
  export = 'export',
  extends = 'extends',
  false = 'false',
  finally = 'finally',
  for = 'for',
  from = 'from',
  function = 'function',
  get = 'get',
  if = 'if',
  implements = 'implements',
  import = 'import',
  in = 'in',
  infer = 'infer',
  instanceof = 'instanceof',
  interface = 'interface',
  keyof = 'keyof',
  let = 'let',
  meta = 'meta',
  module = 'module',
  namespace = 'namespace',
  never = 'never',
  new = 'new',
  null = 'null',
  number = 'number',
  object = 'object',
  of = 'of',
  out = 'out',
  override = 'override',
  private = 'private',
  protected = 'protected',
  public = 'public',
  readonly = 'readonly',
  return = 'return',
  satisfies = 'satisfies',
  set = 'set',
  static = 'static',
  string = 'string',
  super = 'super',
  switch = 'switch',
  symbol = 'symbol',
  this = 'this',
  throw = 'throw',
  true = 'true',
  try = 'try',
  type = 'type',
  typeof = 'typeof',
  undefined = 'undefined',
  unique = 'unique',
  unknown = 'unknown',
  using = 'using',
  var = 'var',
  void = 'void',
  while = 'while',
  with = 'with',
  yield = 'yield'
}

export default keywords
