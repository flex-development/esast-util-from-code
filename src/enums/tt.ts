/**
 * @file Enums - tt
 * @module esast-util-from-code/enums/tt
 */

import type { TokenType } from '@flex-development/vfile-lexer'

/**
 * Token kinds.
 *
 * @see {@linkcode TokenType}
 *
 * @enum {TokenType}
 */
enum tt {
  bigint = 'bigint',
  comment = 'comment',
  eof = 'eof',
  keyid = 'keyid',
  number = 'number',
  punctuator = 'punctuator',
  sof = 'sof',
  string = 'string',
  whitespace = 'whitespace'
}

export default tt
