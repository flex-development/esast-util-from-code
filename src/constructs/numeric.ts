/**
 * @file Constructs - numeric
 * @module esast-util-from-code/constructs/numeric
 */

import { tt } from '#src/enums'
import {
  chars,
  codes,
  ev,
  resolveSlice,
  type Code,
  type Construct,
  type Effects,
  type Event,
  type State,
  type TokenizeContext
} from '@flex-development/vfile-lexer'
import { ok as assert } from 'devlop'
import { asciiDigit } from 'micromark-util-character'

/**
 * Numeric literal construct.
 *
 * @see https://developer.mozilla.org/docs/Web/JavaScript/Reference/Lexical_grammar#numeric_literals
 *
 * @const {Construct} numeric
 */
const numeric: Construct = {
  /**
   * Construct name.
   */
  name: 'numeric',

  /**
   * Resolve the events parsed by `tokenize`.
   *
   * @see {@linkcode Construct.tokenize}
   * @see {@linkcode Event}
   * @see {@linkcode TokenizeContext}
   *
   * @param {Event[]} events - List of events
   * @param {TokenizeContext} context - Tokenize context
   * @return {Event[]} Changed events
   */
  resolve(events: Event[], context: TokenizeContext): Event[] {
    resolveSlice(events, context)

    for (const [event, token] of events) {
      if (event === ev.enter && token.type === tt.number) {
        assert(typeof token.value === 'string', 'expected token value')
        if (token.value.endsWith(chars.lowercaseN)) token.type = tt.bigint
      }
    }

    return events
  },

  /**
   * Check if the current character `code` can start this construct.
   *
   * @see {@linkcode Code}
   * @see {@linkcode TokenizeContext}
   *
   * @this {TokenizeContext}
   *
   * @param {Code} code - Current character code
   * @return {boolean} `true` if `code` can start construct
   */
  test(this: TokenizeContext, code: Code): boolean {
    return asciiDigit(code) || (code === codes.dot && asciiDigit(this.next))
  },

  /**
   * Set up a state machine to handle character codes streaming in.
   *
   * @see {@linkcode Effects}
   * @see {@linkcode State}
   * @see {@linkcode TokenizeContext}
   *
   * @this {TokenizeContext}
   *
   * @param {Effects} effects - Context object to transition state machine
   * @param {State} ok - Successful tokenization state
   * @return {State} Initial state
   */
  tokenize(this: TokenizeContext, effects: Effects, ok: State): State {
    /**
     * Zero check.
     *
     * @const {boolean} zero
     */
    const zero: boolean = this.code === codes.digit0

    /**
     * Map indicating exponential number, where `digit` indicates a digit was
     * seen, and `punctuator` a plus or minus sign was seen.
     *
     * @var {{ digit: boolean; punctuator: boolean }} exponential
     */
    let exponential: { e: boolean; punctuator: boolean } = {
      e: false,
      punctuator: false
    }

    /**
     * Boolean indicating floating point number.
     *
     * @var {boolean} float
     */
    let float: boolean = this.code === codes.dot

    /**
     * Boolean indicating binary, hex, or octal number.
     *
     * @var {boolean} radix
     */
    let radix: boolean = false

    return numeric

    /**
     * Create a radix number state.
     *
     * @param {'b' | 'o' | 'x'} char - Radix number character
     * @return {State} Next state
     */
    function rad(char: 'b' | 'o' | 'x'): State {
      return radix

      /**
       * Finish radix number tokenization.
       *
       * @param {Code} code - Current character code
       * @return {State | undefined} Next state
       */
      function radix(code: Code): State | undefined {
        if (code !== codes.eof) {
          if (code === codes.lowercaseN) return finish(code)

          /**
           * Boolean indicating current code should be consumed.
           *
           * @var {boolean} consume
           */
          let consume: boolean = false

          switch (char) {
            case chars.lowercaseB:
              consume = code >= codes.digit0 && code <= codes.digit1
              break
            case chars.lowercaseO:
              consume = code >= codes.digit0 && code <= codes.digit7
              break
            default:
              consume = asciiDigit(code) ||
                (code >= codes.lowercaseA && code <= codes.lowercaseF) ||
                (code >= codes.uppercaseA && code <= codes.uppercaseF)
              break
          }

          if (!consume && code === codes.underscore) consume = true
          if (consume) return effects.consume(code), radix
        }

        return effects.exit(tt.number), ok(code)
      }
    }

    /**
     * Finish numeric literal tokenization.
     *
     * @param {Code} code - Current character code
     * @return {State | undefined} Next state
     */
    function finish(code: Code): State | undefined {
      if (code === codes.lowercaseN) {
        return effects.consume(code), effects.exit(tt.number), ok
      }

      if (!radix) {
        if (zero) {
          // binary number
          if (code === codes.lowercaseB || code === codes.uppercaseB) {
            return radix = true, effects.consume(code), rad(chars.lowercaseB)
          }

          // hex number
          if (code === codes.lowercaseX || code === codes.uppercaseX) {
            return radix = true, effects.consume(code), rad(chars.lowercaseX)
          }

          // octal number
          if (code === codes.lowercaseO || code === codes.uppercaseO) {
            return radix = true, effects.consume(code), rad(chars.lowercaseO)
          }
        }

        // integer or float
        if (
          asciiDigit(code) ||
          code === codes.underscore ||
          (code === codes.dot && !exponential.e && !float)
        ) {
          if (code === codes.dot) float = true
          return effects.consume(code), finish
        }

        // exponential
        if (code === codes.lowercaseE || code === codes.uppercaseE) {
          if (!exponential.e) {
            return exponential.e = true, effects.consume(code), finish
          }
        }

        // continue exponential
        if (code === codes.minus || code === codes.plus) {
          if (exponential.e && !exponential.punctuator) {
            return exponential.punctuator = true, effects.consume(code), finish
          }
        }
      }

      return effects.exit(tt.number), ok(code)
    }

    /**
     * Start numeric literal tokenization.
     *
     * @param {Code} code - Current character code
     * @return {State} Next state
     */
    function numeric(code: Code): State {
      return effects.enter(tt.number), effects.consume(code), finish
    }
  }
}

export default numeric
