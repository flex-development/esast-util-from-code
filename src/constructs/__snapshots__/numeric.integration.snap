// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:constructs/numeric > bigint > should tokenize bigint (binary) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 bigint "0b11101001010101010101n" (1:1-1:24, 0-23)
│     whitespace: ""
└─2 eof null (1:24-1:24, 23-23)
      whitespace: ""
`;

exports[`integration:constructs/numeric > bigint > should tokenize bigint (exponential) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 bigint "1e3n" (1:1-1:5, 0-4)
│     whitespace: ""
└─2 eof null (1:5-1:5, 4-4)
      whitespace: ""
`;

exports[`integration:constructs/numeric > bigint > should tokenize bigint (float) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 bigint "3.13n" (1:1-1:6, 0-5)
│     whitespace: ""
└─2 eof null (1:6-1:6, 5-5)
      whitespace: ""
`;

exports[`integration:constructs/numeric > bigint > should tokenize bigint (hex) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 bigint "0x123456789ABCDEFn" (1:1-1:19, 0-18)
│     whitespace: ""
└─2 eof null (1:19-1:19, 18-18)
      whitespace: ""
`;

exports[`integration:constructs/numeric > bigint > should tokenize bigint (natural) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 bigint "0n" (1:1-1:3, 0-2)
│     whitespace: ""
└─2 eof null (1:3-1:3, 2-2)
      whitespace: ""
`;

exports[`integration:constructs/numeric > bigint > should tokenize bigint (numeric separators) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 bigint "1_000_000_000_000_000_000_000n" (1:1-1:31, 0-30)
│     whitespace: ""
└─2 eof null (1:31-1:31, 30-30)
      whitespace: ""
`;

exports[`integration:constructs/numeric > bigint > should tokenize bigint (octal) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 bigint "0o777777777777n" (1:1-1:16, 0-15)
│     whitespace: ""
└─2 eof null (1:16-1:16, 15-15)
      whitespace: ""
`;

exports[`integration:constructs/numeric > binary > should tokenize binary number (lowercase) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "0b01010101101" (1:1-1:14, 0-13)
│     whitespace: ""
└─2 eof null (1:14-1:14, 13-13)
      whitespace: ""
`;

exports[`integration:constructs/numeric > binary > should tokenize binary number (numeric separators) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "0b1010_0001" (1:1-1:12, 0-11)
│     whitespace: ""
└─2 eof null (1:12-1:12, 11-11)
      whitespace: ""
`;

exports[`integration:constructs/numeric > binary > should tokenize binary number (uppercase) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "0B11101001010" (1:1-1:14, 0-13)
│     whitespace: ""
└─2 eof null (1:14-1:14, 13-13)
      whitespace: ""
`;

exports[`integration:constructs/numeric > exponential > should tokenize exponential number (float) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "1.3e26" (1:1-1:7, 0-6)
│     whitespace: ""
└─2 eof null (1:7-1:7, 6-6)
      whitespace: ""
`;

exports[`integration:constructs/numeric > exponential > should tokenize exponential number (no punctuators) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "1E3" (1:1-1:4, 0-3)
│     whitespace: ""
└─2 eof null (1:4-1:4, 3-3)
      whitespace: ""
`;

exports[`integration:constructs/numeric > exponential > should tokenize exponential number (numeric separators) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "100_130e-2_6" (1:1-1:13, 0-12)
│     whitespace: ""
└─2 eof null (1:13-1:13, 12-12)
      whitespace: ""
`;

exports[`integration:constructs/numeric > exponential > should tokenize exponential number (signed integer) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "0e+5" (1:1-1:5, 0-4)
│     whitespace: ""
└─2 eof null (1:5-1:5, 4-4)
      whitespace: ""
`;

exports[`integration:constructs/numeric > exponential > should tokenize exponential number (signed integer, negative) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "175e-2" (1:1-1:7, 0-6)
│     whitespace: ""
└─2 eof null (1:7-1:7, 6-6)
      whitespace: ""
`;

exports[`integration:constructs/numeric > float > should tokenize floating point number (leading digit) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "1.3" (1:1-1:4, 0-3)
│     whitespace: ""
└─2 eof null (1:4-1:4, 3-3)
      whitespace: ""
`;

exports[`integration:constructs/numeric > float > should tokenize floating point number (leading digits) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "26.5" (1:1-1:5, 0-4)
│     whitespace: ""
└─2 eof null (1:5-1:5, 4-4)
      whitespace: ""
`;

exports[`integration:constructs/numeric > float > should tokenize floating point number (leading dot) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number ".1398" (1:1-1:6, 0-5)
│     whitespace: ""
└─2 eof null (1:6-1:6, 5-5)
      whitespace: ""
`;

exports[`integration:constructs/numeric > float > should tokenize floating point number (missing tail) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "3." (1:1-1:3, 0-2)
│     whitespace: ""
└─2 eof null (1:3-1:3, 2-2)
      whitespace: ""
`;

exports[`integration:constructs/numeric > float > should tokenize floating point number (numeric separators) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "1_050.95" (1:1-1:9, 0-8)
│     whitespace: ""
└─2 eof null (1:9-1:9, 8-8)
      whitespace: ""
`;

exports[`integration:constructs/numeric > hex > should tokenize hexadecimal number (lowercase) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "0x0123456789FeDcBa" (1:1-1:19, 0-18)
│     whitespace: ""
└─2 eof null (1:19-1:19, 18-18)
      whitespace: ""
`;

exports[`integration:constructs/numeric > hex > should tokenize hexadecimal number (numeric separators) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "0xA0_B0_C0" (1:1-1:11, 0-10)
│     whitespace: ""
└─2 eof null (1:11-1:11, 10-10)
      whitespace: ""
`;

exports[`integration:constructs/numeric > hex > should tokenize hexadecimal number (uppercase) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "0XaBcDeF0123456789" (1:1-1:19, 0-18)
│     whitespace: ""
└─2 eof null (1:19-1:19, 18-18)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (chars.digit0) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "0" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (chars.digit1) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "1" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (chars.digit2) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "2" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (chars.digit3) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "3" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (chars.digit4) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "4" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (chars.digit5) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "5" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (chars.digit6) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "6" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (chars.digit7) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "7" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (chars.digit8) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "8" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (chars.digit9) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "9" (1:1-1:2, 0-1)
│     whitespace: ""
└─2 eof null (1:2-1:2, 1-1)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (digits) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "9876543210" (1:1-1:11, 0-10)
│     whitespace: ""
└─2 eof null (1:11-1:11, 10-10)
      whitespace: ""
`;

exports[`integration:constructs/numeric > integer > should tokenize integer (numeric separators) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "1_000_000_000_000" (1:1-1:18, 0-17)
│     whitespace: ""
└─2 eof null (1:18-1:18, 17-17)
      whitespace: ""
`;

exports[`integration:constructs/numeric > octal > should tokenize octal number (lowercase) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "0o01234567" (1:1-1:11, 0-10)
│     whitespace: ""
└─2 eof null (1:11-1:11, 10-10)
      whitespace: ""
`;

exports[`integration:constructs/numeric > octal > should tokenize octal number (numeric separators) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "0o2_2_5_6" (1:1-1:10, 0-9)
│     whitespace: ""
└─2 eof null (1:10-1:10, 9-9)
      whitespace: ""
`;

exports[`integration:constructs/numeric > octal > should tokenize octal number (uppercase) 1`] = `
tokens[3]
├─0 sof null (1:1-1:1, 0-0)
│     whitespace: ""
├─1 number "0O76543210" (1:1-1:11, 0-10)
│     whitespace: ""
└─2 eof null (1:11-1:11, 10-10)
      whitespace: ""
`;
